tensorflow/compiler/xla/service/llvm_ir/ir_array.h

```diff
@@ -66,6 +66,13 @@ class IrArray {
     // Precondition: "shape" has a layout.
     Index(llvm::Value* linear, const Shape& shape, llvm::IRBuilder<>* b);
 
+    // As before, but also take a multidim to reuse.  multidim.size()
+    // == shape.rank() must be true.  If some of the multidim element
+    // are null we will use the value that would be used if
+    // deliearized from linear.
```

Comment:
"delinearized"?

Could you expand what `multidim` is doing? Is it technically redundant since we already have `shape` and `linear` but there for performance?

Reply:
This constructor is a mix of the constructor at those two:
https://github.com/tensorflow/tensorflow/pull/48381/files/fc5d57766d3b546709d4e6c3e2bdf75f78f2973c#diff-4e1e35c17e8305e625368867ce8e627ace19a9a06bbbf59bdaa96070533e237bR67 (line 67)
https://github.com/tensorflow/tensorflow/pull/48381/files/fc5d57766d3b546709d4e6c3e2bdf75f78f2973c#diff-4e1e35c17e8305e625368867ce8e627ace19a9a06bbbf59bdaa96070533e237bR86 (line 86)

An Index object always have the multidim member. The linear member is optional (see line 199) and is an optimization. I do not want to remove the existing optimization, so I keep linear present.
But I need a simpler multidim value, otherwise LLVM isn't able to simplify the indexing and infer the vectorization.
So I manually changes the inner most multidim value to a simple indexing.

In theory, we could modify LLVM to be smarter about doing the indexing simplification. I did it once in a simple case where I just changed the optimization order. But doing so is time consuming and relying on LLVM to do complex math simplification is a brittle strategy. In this case, I think changing LLVM would request a great changes to its math simplification algo that seems to work well on power of 2, but not on others numbers. And this is what we need, something reliable to vectorize for any shapes values. So I prefer to generates simple indexing. LLVM will be happy. No need for complex changes to LLVM. So it get done faster and is reliable.